### PR TITLE
[Distributed] Handle DA declared inside a generic type

### PR DESCRIPTION
**Pull request by ktoso**
We were not considering the outer scope when forwarding generic substitutions.

Such nesting would not get the generic sub on the distributed call recorded, and thus the call would not be able to be executed on the remote peer:

```swift
struct OuterGeneric<Element>: Sendable, Codable where Element: Sendable & Codable {
  distributed actor Greeter {
    distributed func hello() -> String {
    }
  }
}
```

Resolves rdar://122596546

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/Sema/CodeSynthesisDistributedActor.cpp [hunk 1]
```diff
@@ -346,17 +346,15 @@ deriveBodyDistributed_thunk(AbstractFunctionDecl *thunk, void *context) {
 
   // --- Recording invocation details
   // -- recordGenericSubstitution(s)
-  if (thunk->isGeneric() || nominal->isGeneric()) {
+  if (auto genEnv = thunk->getGenericEnvironment()) {
```
Comment (slavapestov) [inline lib/Sema/CodeSynthesisDistributedActor.cpp:349]: You don't need a GenericEnvironment unless you want to produce archetypes. Maybe you want `if (auto genericSig = thunk->getGenericSignature())`

**Review by slavapestov — APPROVED** [review]

File: lib/Sema/CodeSynthesisDistributedActor.cpp (hunk 1 above)
Comment (ktoso) [reply lib/Sema/CodeSynthesisDistributedActor.cpp:349]: Discussed some more offline, we ned the env -- but we can use it directly below in `mapTypeIntoContext`, I'll add that.

Thanks Slava!

Comment (ktoso) [reply lib/Sema/CodeSynthesisDistributedActor.cpp:349]: done here https://github.com/apple/swift/pull/71492/files#diff-c5ee3c123c92773a4f07b456c2c0051c7a9dfdc81c5c0234ba5da6389bd82a07R358

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Pull request merged**